src/smbprotocol/open.py

```diff
@@ -1523,3 +1602,43 @@ def _close_response(self, request):
             self.end_of_file = c_resp["end_of_file"].get_value()
             self.file_attributes = c_resp["file_attributes"].get_value()
         return c_resp
+
+    def lock(self, locks, lsn=0, lsi=0, wait=True, send=True):
+        """
+        Locks or unlocks byte regions of a file.
+
+        Supports out of band send function, call this function with send=False
+        to return a tuple of (SMB2LockRequest, receive_func) instead of
+        sending the the request and waiting for the response. The receive_func
+        can be used to get the response from the server by passing in the
+        Request that was used to sent it out of band.
+
+        :param locks: (List<SMB2LockElement>) byte ranges to lock or unlock
+        :param lsn: LockSequenceNumber. Only used for SMB dialects > 2.0.2
+        :param lsi: LockSequenceIndex. Only used for SMB dialects > 2.0.2
+        :param wait: If send=True, whether to wait for a response if
+            STATUS_PENDING was received from the server or fail.
+        :param send: Whether to send the request in the same call or return the
+            message to the caller and the unpack function
+        :return: SMB2LockResponse message received from the server
+        """
+
+        lock = SMB2LockRequest()
+        lock["file_id"] = self.file_id
+        lock["locks"] = locks
+
+        if self.connection.dialect > Dialects.SMB_2_0_2:
+            lock["lock_sequence"] = (lsn << 28) + lsi
```

Comment:
If we do decide to expose this, we should ensure that `lsn` fits into the 28 bits and `lsi` isn't more than 4 bits. If they are then we should raise a `ValueError`.

Reply:
Sure. I'll add those checks.